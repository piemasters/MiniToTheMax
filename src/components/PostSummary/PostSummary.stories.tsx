import type { Meta, StoryObj } from '@storybook/react-vite';

import { PostSummary } from './PostSummary';

const meta: Meta<typeof PostSummary> = {
  component: PostSummary,
  argTypes: {
    date: {
      name: 'date',
      description: 'Date the post was published',
      defaultValue: '11/01/2019',
      control: { type: 'date' },
    },
    img: {
      name: 'img',
      description: 'Cover image',
      control: { type: 'object' },
    },
    slug: {
      name: 'slug',
      description: 'The trailing URL of the post',
      defaultValue: 'post-slug',
      control: { type: 'text' },
    },
    title: {
      name: 'title',
      description: 'The title of the post',
      defaultValue: 'Post Title',
      control: { type: 'text' },
    },
    excerpt: {
      name: 'excerpt',
      description: 'The first few lines of the post',
      defaultValue: 'Post Title',
      control: { type: 'text' },
    },
    tall: {
      name: 'tall',
      description: 'Increase the offset of the title for taller images',
      defaultValue: false,
      control: { type: 'boolean' },
    },
  },
};

export default meta;
type Story = StoryObj<typeof PostSummary>;

export const Primary: Story = {
  args: {
    date: '11/01/2019',
    img: {
      layout: 'constrained', // "fullWidth" | "fixed";
      backgroundColor: '#585858',
      width: 1800,
      height: 900,
      images: {
        fallback: {
          src: '/images/example-cover.jpg',
          srcSet:
            '/images/example-cover.jpg 450w,\n/images/example-cover.jpg 900w,\n/images/example-cover.jpg 1800w',
          sizes: '(min-width: 1800px) 1800px, 100vw',
        },
        sources: [
          {
            srcSet:
              '/images/example-cover.jpg 450w,\n/images/example-cover.jpg 900w,\n/images/example-cover.jpg 1800w',
            type: 'image/jpg',
            sizes: '(min-width: 1800px) 1800px, 100vw',
          },
        ],
      },
    },
    slug: 'post-slug',
    title: 'Post Title',
    excerpt: 'The is the excerpt of the post',
    tall: true,
  },
};
