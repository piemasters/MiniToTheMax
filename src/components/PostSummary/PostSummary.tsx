import React, { FC } from 'react';
import type { IGatsbyImageData } from 'gatsby-plugin-image';

import { CoverImage } from '../CoverImage';
import { PageLink } from '../PageLink';

export interface PostSummaryProps {
  slug: string;
  title: string;
  date: string;
  img?: IGatsbyImageData;
  excerpt: string;
  tall?: boolean;
}

export const PostSummary: FC<PostSummaryProps> = ({
  slug,
  title,
  date,
  img,
  excerpt,
  tall,
}) => {
  return (
    <div className="my-4" data-testid="post-summary">
      <PageLink
        className="block no-underline bg-gray-200 duration-300 hover:shadow-3xl"
        to={slug}
        type={'cover'}
        direction={'up'}
      >
        {img && <CoverImage image={img} title={title} tall={tall} />}
        <div className="p-4 text-gray-600">
          <p className="text-xs font-bold mb-2">{date}</p>
          <hr />
          <p className="italic m-0 text-xs/loose text-gray-600">{excerpt}</p>
        </div>
      </PageLink>
    </div>
  );
};

export default PostSummary;
