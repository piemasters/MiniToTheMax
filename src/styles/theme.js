export default {
  colors: {
    text: '#000',
    muted: '#999',
  },
};
