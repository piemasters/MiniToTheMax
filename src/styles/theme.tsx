export interface Theme {
  colors: {
    text: string;
    muted: string;
  };
}

const appTheme: Theme = {
  colors: {
    text: '#000',
    muted: '#999',
  },
};

export default appTheme;
