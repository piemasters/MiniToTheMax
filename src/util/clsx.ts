/* eslint-disable @typescript-eslint/no-unused-expressions */
/* eslint-disable @typescript-eslint/no-explicit-any */
export type ClassValue =
  | ClassArray
  | ClassDictionary
  | string
  | number
  | null
  | boolean
  | undefined;
type ClassDictionary = Record<string, any>;
type ClassArray = ClassValue[];

function toVal(mix: string | number | boolean | unknown[] | ClassDictionary) {
  let k,
    y,
    str = '';

  if (typeof mix === 'string' || typeof mix === 'number') {
    str += mix;
  } else if (typeof mix === 'object') {
    if (Array.isArray(mix)) {
      const len = mix.length;
      for (k = 0; k < len; k++) {
        if (mix[k]) {
          if ((y = toVal(mix[k]))) {
            str && (str += ' ');
            str += y;
          }
        }
      }
    } else {
      for (y in mix) {
        if (mix[y]) {
          str && (str += ' ');
          str += y;
        }
      }
    }
  }

  return str;
}

export function clsx(...args: ClassValue[]): string {
  let i = 0,
    tmp,
    x,
    str = '';
  const len = args.length;
  for (; i < len; i++) {
    if ((tmp = args[i])) {
      if ((x = toVal(tmp))) {
        str && (str += ' ');
        str += x;
      }
    }
  }
  return str;
}

export default clsx;
