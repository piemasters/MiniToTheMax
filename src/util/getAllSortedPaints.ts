import { PaintDetails } from '../types';
import { textToId } from './textToId';
import { sortPaints } from './sortPaints';
import {
  airPaints,
  basePaints,
  contrastPaints,
  dryPaints,
  layerPaints,
  shadePaints,
  sprayPaints,
  technicalPaints,
} from '../data/paints/citadel';
import { gameColorPaints } from '../data/paints';

export const getAllSortedPaints = (): PaintDetails[] => {
  const allPaints: PaintDetails[] = [
    // citadel
    ...airPaints,
    ...basePaints,
    ...contrastPaints,
    ...dryPaints,
    ...layerPaints,
    ...shadePaints,
    ...sprayPaints,
    ...technicalPaints,
    // vallejo
    ...gameColorPaints,
  ];

  // Remove duplicates using a Set
  const uniquePaints = Array.from(
    new Set(
      allPaints.map(
        (paint) =>
          `${textToId(paint.company)}_${textToId(paint.name)}_${textToId(paint.category)}`
      )
    )
  ).map((uniqueKey) =>
    allPaints.find(
      (paint) =>
        `${textToId(paint.company)}_${textToId(paint.name)}_${textToId(paint.category)}` ===
        uniqueKey
    )
  ) as PaintDetails[];

  // Sort the paints using the sortPaints function
  return uniquePaints.sort(sortPaints);
};
