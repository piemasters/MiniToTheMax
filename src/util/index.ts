export * from './pagination';
export * from './paintFilters';
