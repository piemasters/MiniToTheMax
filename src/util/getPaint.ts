import {
  CitadelPaintCategories,
  PaintCategory,
  PaintDetails,
  VallejoPaintCategories,
} from '../types';
import { getAllSortedPaints } from './getAllSortedPaints';

const allPaints = getAllSortedPaints();

const paintCategoryPriority = [
  ...CitadelPaintCategories,
  ...VallejoPaintCategories,
] as const;

type category = (typeof paintCategoryPriority)[number];

export const getPaint = (
  name: string,
  category?: PaintCategory
): { paint?: PaintDetails; error?: string; type?: string } => {
  const paintMatches = allPaints.filter((paint) => paint.name === name);

  if (paintMatches.length === 0) {
    return { error: 'Incorrect Paint Name', type: name };
  }

  if (paintMatches.length === 1) {
    return { paint: paintMatches[0] };
  }

  if (category) {
    const categoryMatches = paintMatches.filter(
      (paint) => paint.category === category
    );
    if (categoryMatches.length === 0) {
      return { error: 'Incorrect Paint Category', type: category };
    }
    return { paint: categoryMatches[0] };
  }

  const getCategoryPriority = (category: PaintCategory) => {
    const index = paintCategoryPriority.indexOf(category);
    return index === -1 ? Infinity : index;
  };
  paintMatches.sort(
    (a, b) => getCategoryPriority(a.category) - getCategoryPriority(b.category)
  );

  return { paint: paintMatches[0] };
};
