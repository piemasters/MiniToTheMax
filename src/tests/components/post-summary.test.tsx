import React from 'react';
import PostSummary from '../../components/post-summary';
import { getByTestId } from '@testing-library/react';
import { ThemeProvider } from '@emotion/react';
import theme from '../../styles/theme';
import { renderWithTransitionProvider } from '../util/transition-provider';

describe('PostSummary', () => {
  test('renders correctly', () => {
    const { container } = renderWithTransitionProvider(
      <ThemeProvider theme={theme}>
        <PostSummary
          date={'11/01/2019'}
          img={{
            layout: 'constrained',
            width: 1800,
            height: 900,
            backgroundColor: '#282828',
            images: {
              fallback: {
                src: 'cover.jpeg',
              },
            },
          }}
          slug={'slug'}
          title={'Demo Post Title'}
          excerpt={'A test sentence'}
          timeToRead={5}
        />
      </ThemeProvider>
    );

    expect(
      getByTestId(container as HTMLElement, 'post-summary')
    ).toHaveTextContent('Demo Post Title');
    expect(container).toMatchSnapshot();
  });
});
