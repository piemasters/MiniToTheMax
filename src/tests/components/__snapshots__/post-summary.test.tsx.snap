// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PostSummary renders correctly 1`] = `
<div>
  <div
    css="You have tried to stringify object returned from \`css\` function. It isn't supposed to be used directly (e.g. as value of the \`className\` prop), but rather handed to emotion so it can handle it (e.g. as value of \`css\` prop)."
    data-testid="post-summary"
  >
    <a
      bg="#eb1d23"
      css="You have tried to stringify object returned from \`css\` function. It isn't supposed to be used directly (e.g. as value of the \`className\` prop), but rather handed to emotion so it can handle it (e.g. as value of \`css\` prop)."
      data-testid="page-link-cover"
      direction="up"
      duration="1"
      href="/slug"
    >
      <div
        css="You have tried to stringify object returned from \`css\` function. It isn't supposed to be used directly (e.g. as value of the \`className\` prop), but rather handed to emotion so it can handle it (e.g. as value of \`css\` prop)."
        data-testid="cover-image"
      >
        <div
          class="gatsby-image-wrapper gatsby-image-wrapper-constrained"
          data-gatsby-image-wrapper=""
          style="position: relative; overflow: hidden; display: inline-block;"
        >
          <div
            style="max-width: 1800px; display: block;"
          >
            <img
              alt=""
              aria-hidden="true"
              role="presentation"
              src="data:image/svg+xml;charset=utf-8,%3Csvg height='900' width='1800' xmlns='http://www.w3.org/2000/svg' version='1.1'%3E%3C/svg%3E"
              style="max-width: 100%; display: block; position: static;"
            />
          </div>
          <div
            aria-hidden="true"
            data-placeholder-image=""
            style="height: 100%; left: 0px; position: absolute; top: 0px; width: 100%;"
          />
          <img
            alt="test"
            css="You have tried to stringify object returned from \`css\` function. It isn't supposed to be used directly (e.g. as value of the \`className\` prop), but rather handed to emotion so it can handle it (e.g. as value of \`css\` prop)."
            data-gatsby-image-ssr=""
            data-main-image=""
            data-src="cover.jpeg"
            decoding="async"
            loading="lazy"
            style="height: 100%; left: 0px; position: absolute; top: 0px; transform: translateZ(0); transition: opacity 250ms linear; width: 100%; will-change: opacity; opacity: 0;"
          />
          <noscript />
          <script
            type="module"
          >
            const t="undefined"!=typeof HTMLImageElement&&"loading"in HTMLImageElement.prototype;if(t){const t=document.querySelectorAll("img[data-main-image]");for(let e of t)e.dataset.src&&(e.setAttribute("src",e.dataset.src),e.removeAttribute("data-src")),e.dataset.srcset&&(e.setAttribute("srcset",e.dataset.srcset),e.removeAttribute("data-srcset")),e.complete&&(e.style.opacity=1)}
          </script>
        </div>
        <h2
          css="You have tried to stringify object returned from \`css\` function. It isn't supposed to be used directly (e.g. as value of the \`className\` prop), but rather handed to emotion so it can handle it (e.g. as value of \`css\` prop)."
        >
          Demo Post Title
        </h2>
      </div>
      <div
        css="You have tried to stringify object returned from \`css\` function. It isn't supposed to be used directly (e.g. as value of the \`className\` prop), but rather handed to emotion so it can handle it (e.g. as value of \`css\` prop)."
      >
         
        <p
          css="You have tried to stringify object returned from \`css\` function. It isn't supposed to be used directly (e.g. as value of the \`className\` prop), but rather handed to emotion so it can handle it (e.g. as value of \`css\` prop)."
        >
          <strong>
            11/01/2019
             - 
            5
             min read
          </strong>
        </p>
        <hr />
        <p
          css="You have tried to stringify object returned from \`css\` function. It isn't supposed to be used directly (e.g. as value of the \`className\` prop), but rather handed to emotion so it can handle it (e.g. as value of \`css\` prop)."
        >
          A test sentence
        </p>
      </div>
    </a>
  </div>
</div>
`;
